### ISSUE TITLE
Simulation time increase when using newer version

### ISSUE THREAD

**Issue by quintiminator** [cocotb/cocotb#3111]
Hello guys, here is an update on this topic (original message below the line). I found something interesting on the trigger functions in cocotb I think.

I used the provided cocotb Matrix Multiplier example to test my hypothesis with. SO the hypothesis is that the newer cocotb version 1.7 is slower (and most likely due to either scheduling or the trigger functions). 

When I run the Matrix Multiplier example with only 3000 input data points, they all are finished within a minute. For my projects, I want to test more than "only" 3000 samples. Furthermore, an input sample should not be generated every clock cycle (but every 100 cycles e.g.).

So when running 100000 samples with the Matrix Multiplier, the cocotb version 1.7.1 returned a simulation time which is consistently visibly higher than cocotb version 1.5.1. So the effect on a small model such as the matrix multiplier is only visible when taking more samples.

However, in a larger project that I am working on, the effect of the longer simulation time is visible earlier. When only using 300 input data points for my own project, 1.5.1 would return around 1.5 minutes of simulation time whereas version 1.7.1 returns 1.5 hours with no other changes except the cocotb version.

I tested what it could be by stripping down my test and literally only keep a counter which is increased every cycle. This was already quite slow (by making use of the RisingEdge function). So this made me believe that it has something to do with trigger.py. Unfortunately, I cannot find something drastically different between 1.5 and 1.7 which could cause this. 

So could it be the interface between Python and the simulator? Because checking every X cycles (or even every cycle) if there is a rising edge could take much time. What in 1.7 could cause this slow-down compared to 1.6 and before?
___________________________________________________________________________________________________________________________________
Original problem:
Hello, I am running with Python 3.7 on the 32-bit simulator version.

I was running a rather short simulation with cocotb version 1.5.1 and it fairly quick. However, when comparing the simulation times with 1.6.2 and 1.7.1 there is a significant slowdown.

The 1.5.1 version would be around 200s whereas the 1.6.2 and 1.7.1 result in a simulation time which is around 600 and 900 seconds, respectively.

In another simulation which takes longer, using cocotb 1.5.1. resulted in a simulation time less than 1000 seconds whereas the newer versions could not even finish within hours.

Besides the cocotb version, there were no other variables in this test as they were all run in a docker container. I assume this slowdown has probably something to do with the scheduler but I cannot find significant changes between 1.5.X and 1.6.X and 1.7.X when comparing sources codes.

What else could have such a great influence on the time? Is it connected to how the new coroutines are planned? the fork vs start_soon  for example? However, I do not see how this would be affect the simulation this greatly. The Python version remained the same.

___________________________________________________________________________________________________________________________________

**Comment by felixdube** [comment]
You might want to run the profiler on both version and compare.
https://docs.cocotb.org/en/stable/building.html?highlight=profiling#envvar-COCOTB_ENABLE_PROFILING

**Comment by quintiminator** [comment]
Hello, thank you for the reply!

I actually did the profiling but the graphs did not make much sense to me.

It seemed all relative and not absolute(?). The scheduler e..g. was in all three versions (tested 1.5, 1.6 and 1.7) 97% and the 'subtrees' were all similar. It did not show the exact time. And also the decorator and monitor class were in all three cases similar percentage. Where can I find how to read such graph? I used graph2dot.

**Comment by ktbarrett** [comment]
Can you post the results here? They do require a lot of interpretation.

**Comment by themperek** [comment]
BTW: One could use this to get line profiler (maybe easier to read) : https://github.com/vpelletier/pprofile

**Comment by quintiminator** [comment]
@themperek  I will take a look at it, from the doc alone it seems intuitive!

@ktbarrett for example a snippet of 1.5:
![image](https://user-images.githubusercontent.com/35113959/196424087-987e7191-d5b5-4d29-b9b4-6394eba00a13.png)

Snippet of 1.7:
![image](https://user-images.githubusercontent.com/35113959/196423980-64e3a1c1-a1cb-4f2d-8abb-ba4acb4be4ff.png)

**Comment by themperek** [comment]
Maybe here: https://github.com/cocotb/cocotb/blob/77e22580bb189bdaf4367d7fa6aecba6c2505a77/cocotb/scheduler.py#L331

```python3
    import pprofile
    prof = pprofile.Profile()
    with prof():

      # code

    prof.print_stats()
```

instead od `ctx`. Or similar,

**Comment by quintiminator** [comment]
Hello, so I tried some profiling but it was really difficult with the asynchronous set-up and cocotb framework around it.

I tried finding the problem in a different manner and I might have found what causes the simulation time to be drastically greater. I updated my initial comment.

But basically, I think it could have something to do with the communication between the Python framework and the simulator such as Modelsim when checking for rising edges (or wait for X cycles). I checked this by removing all features and only kept a counter which increases every clock cycle. So I used a RisingEdge function (and even await Timer) to test how fast it is, and it is much slower in 1.7 than in 1.5.

**Comment by ktbarrett** [comment]
@quintiminator What version of Modelsim are you using?

**Comment by quintiminator** [comment]
@ktbarrett I did not expect response anymore because I posted it a long time ago. Anyway, I am using Modelsim version 2019.3. I will read the posts of #3229 first because that person has the same issue.

**Comment by ktbarrett** [comment]
@quintiminator If the issue has been resolved with the development version of cocotb, consider closing this issue.

**Comment by quintiminator** [comment]
@ktbarrett  I am currently checking, I will close it when it is solved!

**Comment by quintiminator** [comment]
It is indeed fixed in the 1.8.0 development version!
